plugins/languages/treesitter/treesitter.nix

```
@@ -6,210 +6,264 @@
-      package = mkOption {
-        type = types.package;
-        default = pkgs.vimPlugins.nvim-treesitter;
-        description = "Plugin to use for nvim-treesitter. If using nixGrammars, it should include a `withPlugins` function";
-      };
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "treesitter";
+  originalName = "nvim-treesitter";
+  luaName = "nvim-treesitter.configs";
+  defaultPackage = pkgs.vimPlugins.nvim-treesitter;
 
-      nixGrammars = mkOption {
-        type = types.bool;
-        default = true;
-        description = "Install grammars with Nix";
-      };
+  maintainers = [ ];
 
-      ensureInstalled = mkOption {
-        type =
-          with types;
-          oneOf [
-            (enum [ "all" ])
-            (listOf str)
-          ];
-        default = "all";
-        description = "Either \"all\" or a list of languages";
-      };
+  optionsRenamedToSettings = [
+    "ensureInstalled"
+    "ignoreInstall"
+    "indent"
+    "parserInstallDir"
+    [
+      "incrementalSelection"
+      "enable"
+    ]
+    [
+      "incrementalSelection"
+      "keymaps"
+      "initSelection"
+      "nodeDecremental"
+    ]
+    [
+      "incrementalSelection"
+      "keymaps"
+      "initSelection"
+      "nodeIncremental"
+    ]
+    [
+      "incrementalSelection"
+      "keymaps"
+      "initSelection"
+      "scopeIncremental"
+    ]
+  ];
 
-      gccPackage = helpers.mkPackageOption {
-        default = if cfg.nixGrammars then null else pkgs.gcc;
-        description = ''
-          Which package (if any) to be added as the GCC compiler.
-          This is required to build grammars if you are not using `nixGrammars`.
-          To disable the installation of GCC, set this option to `null`.
-        '';
-      };
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "treesitter"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (lib.mkRenamedOptionModule (basePluginPath ++ [ "customCaptures" ]) (
+        settingsPath
+        ++ [
+          "highlight"
+          "custom_captures"
+        ]
+      ))
+      (lib.mkRenamedOptionModule (basePluginPath ++ [ "disabledLanguages" ]) (
+        settingsPath
+        ++ [
+          "highlight"
+          "disable"
+        ]
+      ))
+    ];
 
-      parserInstallDir = mkOption {
-        type = types.nullOr types.str;
-        default = if cfg.nixGrammars then null else "$XDG_DATA_HOME/nvim/treesitter";
-        description = ''
-          Location of the parsers to be installed by the plugin (only needed when nixGrammars is disabled).
-          This default might not work on your own install, please make sure that $XDG_DATA_HOME is set if you want to use the default. Otherwise, change it to something that will work for you!
-        '';
-      };
+  settingsOptions = {
+    additional_vim_regex_highlighting =
+      with types;
+      helpers.defaultNullOpts.mkNullable (either bool (listOf str)) false ''
```

```suggestion
      with helpers.nixvimTypes;
      helpers.defaultNullOpts.mkNullableWithRaw (either bool (listOf (maybeRaw str))) false ''
```

`mkListOf` would wrap the elemType with `maybeRaw`. We should replicate that as it allows users to set list items to raw lua code (e.g. use a string defined in a local variable somewhere)

mkNullableWithRaw allows to option to be rawLua at it's top-level. `mkListOf` doesn't do that currently, but that's a bug.

You could also do this if preferred, it's effectively the same thing:

```suggestion
      with helpers.nixvimTypes;
      helpers.defaultNullOpts.mkNullable (oneOf [
        bool
        (listOf (maybeRaw str))
        rawLua
      ]) false ''
```